go/teams/audit.go

```diff
@@ -1,15 +1,409 @@
+		if prev != nil && prev.team > tuple.team {
+			return maxMerkleProbe, NewAuditError("team chain unexpected jump: %d > %d via merkle seqno %d", prev.team, tuple.team, tuple.merkle)
+		}
+		if tuple.merkle > maxMerkleProbe {
+			maxMerkleProbe = tuple.merkle
+		}
+		prev = &tuple
+	}
+	return maxMerkleProbe, nil
+}
+
+func (a *Auditor) doPreProbes(m libkb.MetaContext, history *keybase1.AuditHistory, probeID int, headMerkle keybase1.MerkleRootV2) (err error) {
+	defer m.CTrace("Auditor#doPreProbes", func() error { return err })()
+
+	first := m.G().MerkleClient.FirstSeqnoWithSkips(m)
+	if first == nil {
+		return NewAuditError("cannot find a first modern merkle sequence")
+	}
+
+	probeTuples, err := a.computeProbes(m, history.ID, history.PreProbes, probeID, *first, headMerkle.Seqno, len(history.PreProbes), params.NumPreProbes)
+	if err != nil {
+		return err
+	}
+	if len(probeTuples) == 0 {
+		m.CDebugf("No probe pairs, so bailing")
+		return nil
+	}
+	for _, tuple := range probeTuples {
+		m.CDebugf("preProbe: checking probe at merkle %d", tuple.merkle)
+		if tuple.team != keybase1.Seqno(0) || !tuple.linkID.IsNil() {
+			return NewAuditError("merkle root at %v should have been nil for %v; got %s/%d",
+				tuple.merkle, history.ID, tuple.linkID, tuple.team)
+		}
+	}
+	return nil
+}
+
+func randSeqno(lo keybase1.Seqno, hi keybase1.Seqno) (keybase1.Seqno, error) {
+	rng := hi - lo + 1
+	rngBig := big.NewInt(int64(rng))
+	n, err := rand.Int(rand.Reader, rngBig)
+	if err != nil {
+		return keybase1.Seqno(0), err
+	}
+	return keybase1.Seqno(n.Int64()) + lo, nil
+}
+
+type probeTuple struct {
+	merkle keybase1.Seqno
+	team   keybase1.Seqno
+	linkID keybase1.LinkID
+}
+
+func (a *Auditor) computeProbes(m libkb.MetaContext, teamID keybase1.TeamID, probes map[keybase1.Seqno]keybase1.Probe, probeID int, left keybase1.Seqno, right keybase1.Seqno, probesInRange int, n int) (ret []probeTuple, err error) {
+	ret, err = a.scheduleProbes(m, probes, probeID, left, right, probesInRange, n)
+	if err != nil {
+		return nil, err
+	}
+	err = a.lookupProbes(m, teamID, ret)
+	if err != nil {
+		return nil, err
+	}
+	return ret, err
+}
+
+func (a *Auditor) scheduleProbes(m libkb.MetaContext, probes map[keybase1.Seqno]keybase1.Probe, probeID int, left keybase1.Seqno, right keybase1.Seqno, probesInRange int, n int) (ret []probeTuple, err error) {
+	defer m.CTrace(fmt.Sprintf("Auditor#scheduleProbes(left=%d,right=%d)", left, right), func() error { return err })()
+	if probesInRange > n {
+		m.CDebugf("no more probes needed; did %d, wanted %d", probesInRange, n)
+		return nil, nil
+	}
+	rng := right - left + 1
+	if int(rng) <= probesInRange {
+		m.CDebugf("no more probes needed; range was only %d, and we did %d", rng, probesInRange)
+		return nil, nil
+	}
+	for i := 0; i < n; i++ {
+		x, err := randSeqno(left, right)
+		if err != nil {
+			return nil, err
+		}
+		if _, found := probes[x]; !found {
```

Comment:
This isn't a huge deal of course but it's expected more like 24.7 distincts instead of 25. Should be careful not to reduce the MerkleMovementTrigger too much or we might end up getting significantly less (at 50, only 20/25 expected).